FirebaseAdmin/FirebaseAdmin/Messaging/InstanceIdClient.cs

```diff
@@ -0,0 +1,181 @@
+    /// Initializes a new instance of the <see cref="InstanceIdClient"/> class.
+    /// </summary>
+    /// <param name="clientFactory">A default implentation of the HTTP client factory.</param>
+    /// <param name="credential">An instance of the <see cref="GoogleCredential"/> GoogleCredential class.</param>
+    /// <param name="projectId">The Project Id for FCM Messaging.</param>
+    public InstanceIdClient(HttpClientFactory clientFactory, GoogleCredential credential, string projectId)
+    {
+        if (string.IsNullOrEmpty(projectId))
+        {
+            throw new ArgumentException(
+                "Project ID is required to access messaging service. Use a service account "
+                + "credential or set the project ID explicitly via AppOptions. Alternatively "
+                + "you can set the project ID via the GOOGLE_CLOUD_PROJECT environment "
+                + "variable.");
+        }
+
+        this.httpClient = clientFactory.ThrowIfNull(nameof(clientFactory))
+            .CreateAuthorizedHttpClient(credential);
+
+        this.errorHandler = new MessagingErrorHandler();
+    }
+
+    /// <summary>
+    /// Subscribes a list of registration tokens to a topic.
+    /// </summary>
+    /// <param name="topic">The topic name to subscribe to. /topics/ will be prepended to the topic name provided if absent.</param>
+    /// <param name="registrationTokens">A list of registration tokens to subscribe.</param>
+    /// <returns>A task that completes with a <see cref="TopicManagementResponse"/>, giving details about the topic subscription operations.</returns>
+    public async Task<TopicManagementResponse> SubscribeToTopicAsync(string topic, List<string> registrationTokens)
+    {
+        try
+        {
+            return await this.SendInstanceIdRequest(topic, registrationTokens, IidSubscriberPath).ConfigureAwait(false);
+        }
+        catch (HttpRequestException e)
+        {
+            throw this.CreateExceptionFromResponse(e);
+        }
+        catch (IOException)
+        {
+            throw new FirebaseMessagingException(ErrorCode.Internal, "Error while calling IID backend service");
+        }
+    }
+
+    /// <summary>
+    /// Unsubscribes a list of registration tokens from a topic.
+    /// </summary>
+    /// <param name="topic">The topic name to unsubscribe from. /topics/ will be prepended to the topic name provided if absent.</param>
+    /// <param name="registrationTokens">A list of registration tokens to unsubscribe.</param>
+    /// <returns>A task that completes with a <see cref="TopicManagementResponse"/>, giving details about the topic unsubscription operations.</returns>
+    public async Task<TopicManagementResponse> UnsubscribeFromTopicAsync(string topic, List<string> registrationTokens)
+    {
+        try
+        {
+            return await this.SendInstanceIdRequest(topic, registrationTokens, IidUnsubscribePath).ConfigureAwait(false);
+        }
+        catch (HttpRequestException e)
```

Comment:
SendInstancedIdRequest already does the error handling. This shouldn't be needed.